src/compiler/factory.ts

```diff
@@ -122,4 +122,210 @@ namespace ts {
         block.statements = createNodeArray(statements);
         return block;
     }
+    export function createVariableDeclaration(name: BindingPattern | Identifier, initializer?: Expression, location?: TextRange): VariableDeclaration {
+        const node = <VariableDeclaration>createNode(SyntaxKind.VariableDeclaration, location);
+        node.name = name;
+        node.initializer = initializer;
+        return node;
+    }
+
+    export function createIdentifier(text: string): Identifier {
+        const node = <Identifier>createNode(SyntaxKind.Identifier);
+        node.text = text;
+        return node;
+    }
+
+    export function createTempVariable(tempKind: TempVariableKind): Identifier {
+        const name = <Identifier>createNode(SyntaxKind.Identifier);
+        name.tempKind = tempKind;
+        getNodeId(name);
+        return name;
+    }
+
+    export function createLiteral(value: string): StringLiteral;
+    export function createLiteral(value: number): LiteralExpression;
+    export function createLiteral(value: string | number | boolean): PrimaryExpression;
+    export function createLiteral<T extends PrimaryExpression>(value: string | number | boolean): T {
+        if (typeof value === "number") {
+            const node = <T & LiteralExpression>createNode(SyntaxKind.NumericLiteral);
+            node.text = value.toString();
+            return node;
+        }
+        else if (typeof value === "boolean") {
+            return <T>createNode(value ? SyntaxKind.TrueKeyword : SyntaxKind.FalseKeyword);
+        }
+        else {
+            const node = <T & StringLiteral>createNode(SyntaxKind.StringLiteral);
+            node.text = String(value);
+            return node;
+        }
+    }
+
+    export function createVoid(expression: UnaryExpression) {
+        const node = <VoidExpression>createNode(SyntaxKind.VoidExpression);
+        node.expression = expression;
+        return node;
+    }
+
+    export function createVoidZero() {
+        return createVoid(createLiteral(0));

```

Comment:
My concern with caching these expressions are cases where we may want to supply a position for the node to assist in source maps. We don't want to accidentally apply a position to a node that's in use elsewhere. This also becomes a concern if we eventually open up transformations to third parties at any point.